### PR TITLE
Added solution to usaco-1040

### PR DESCRIPTION
**Pull request by buffering07**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

**Comment by ryanchou-dev** [comment]
err, yeah, same suggestions | #2746 

you don't have to close a pr to edit it...

File: solutions/usaco-1040.mdx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		for (int i = 0; i < n; i++) {
+			x.add(io.nextInt());
+			y.add(io.nextInt());
+			components.add(i);
+		}
+								 
+		Collections.sort(components, new Comparator<Integer> () {         
+			public int compare(Integer p1, Integer p2) // sorts by x, then y {
+				if (x.get(p1).equals(x.get(p2)))
+				{
```
Comment (ryanchou-dev) [inline solutions/usaco-1040.mdx:46]: .

File: solutions/usaco-1040.mdx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+**Time Complexity:** $\mathcal{O}(N \log N)$
+
+<LanguageSection>
+	
+<JavaSection>
+	
+```java
+import java.util.*;
+import java.io.*;
+public class TheMooParticle {
```
Comment (ryanchou-dev) [inline solutions/usaco-1040.mdx:28]:
```suggestion

public class TheMooParticle {
```

File: solutions/usaco-1040.mdx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio("moop");
+		int n = io.nextInt(); // number of particles
+	
+		ArrayList<Integer> x = new ArrayList<>(); // x coordinates
+		ArrayList<Integer> y = new ArrayList<>(); // y coordinates
```
Comment (ryanchou-dev) [inline solutions/usaco-1040.mdx:34]: inline comments should be separated by two spaces

```
code  // comment
```

File: solutions/usaco-1040.mdx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		maxRight[n - 1] = y.get(components.get(n - 1));
+		for (int i = n - 2; i >= 0; i--) {
+			// calculates maxRight using suffix array
+			maxRight[i] = Math.max(maxRight[i + 1], y.get(components.get(i))); 
+		}
+		int numComponents = 1;
+		for (int i = 0; i < n - 1; i++) {
+			// means they are not in the same component
+			if (minLeft[i] > maxRight[i + 1]) 
+			{
```
Comment (ryanchou-dev) [inline solutions/usaco-1040.mdx:69]: .

File: solutions/usaco-1040.mdx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		for (int i = 1; i < n; i++) {
+			// calculates minLeft using prefix array
+			minLeft[i] = Math.min(minLeft[i - 1], y.get(components.get(i))); 
+		}
+		maxRight[n - 1] = y.get(components.get(n - 1));
+		for (int i = n - 2; i >= 0; i--) {
+			// calculates maxRight using suffix array
+			maxRight[i] = Math.max(maxRight[i + 1], y.get(components.get(i))); 
+		}
+		int numComponents = 1;
```
Comment (ryanchou-dev) [inline solutions/usaco-1040.mdx:65]: space this out a bit, it's too hard to read

**Review by ryanchou-dev — APPROVED** [review]
🎄

**Review by ryanchou-dev — APPROVED** [review]
yep! thanks for helping out -w-

**Comment by buffering07** [comment]
thank you!

**Review by jessechoe10 — APPROVED** [review]

File: solutions/usaco-1040.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-1040.mdx:29]: what this for

File: solutions/usaco-1040.mdx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.*;
+import java.io.*;
+	
+public class TheMooParticle {
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio("moop");
+		int n = io.nextInt();  // number of particles
+	
+		ArrayList<Integer> x = new ArrayList<>();  // x coordinates
+		ArrayList<Integer> y = new ArrayList<>();  // y coordinates
```
Comment (SansPapyrus683) [inline solutions/usaco-1040.mdx:35]: comments need not state the obvious

Comment (SansPapyrus683) [inline solutions/usaco-1040.mdx:35]:
```suggestion
		List<Integer> x = new ArrayList<>();
		List<Integer> y = new ArrayList<>();
```

File: solutions/usaco-1040.mdx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		ArrayList<Integer> components = new ArrayList<>();
```
Comment (SansPapyrus683) [inline solutions/usaco-1040.mdx:36]: change the arraylist dec to list

File: solutions/usaco-1040.mdx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+				if (x.get(p1).equals(x.get(p2)))
+				{
+					return y.get(p1).compareTo(y.get(p2));
+				}
+				return x.get(p1).compareTo(x.get(p2));
+			}
+		});
+	
+		int[] minLeft = new int[100000];  // stores min y-coordinate for left comp
+		int[] maxRight = new int[100000];  // stores max y-coordinate for right comp
```
Comment (SansPapyrus683) [inline solutions/usaco-1040.mdx:55]: put 1e5 in a magic number like MAX_POS or something

File: solutions/usaco-1040.mdx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+			if (minLeft[i] > maxRight[i + 1]) 
+			{
+				numComponents++;
+			}
+		}
+	
+		io.println(numComponents);
+		io.close();
+		
+
```
Comment (SansPapyrus683) [inline solutions/usaco-1040.mdx:80]:
```suggestion
```

File: solutions/usaco-1040.mdx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+				numComponents++;
+			}
+		}
+	
+		io.println(numComponents);
+		io.close();
+		
+		
+	}
+	//CodeSnip{Kattio}
```
Comment (SansPapyrus683) [inline solutions/usaco-1040.mdx:82]:
```suggestion

	//CodeSnip{Kattio}
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**